### DISCUSSION TITLE
Cannot see ISO images?

### DISCUSSION THREAD

**Discussion by c2b2pss** [MusicPlayerDaemon/MPD#1415] [Q&A]
Everything else works and music plays from all other file including DSD.

But I cannot "see" SACD ISO images ......I am using a simple frontend GMPC as trial 

Any ideas please? Thanks!!

```
Music Player Daemon 0.24 (0.24)
Copyright 2003-2007 Warren Dukes <warren.dukes@gmail.com>
Copyright 2008-2021 Max Kellermann <max.kellermann@gmail.com>
This is free software; see the source for copying conditions.  There is NO
warranty; not even MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Database plugins:
 simple proxy upnp

Storage plugins:
 local udisks nfs curl

Neighbor plugins:
 upnp udisks

Decoders plugins:
 [mad] mp3 mp2
 [mpg123] mp3
 [vorbis] ogg oga
 [oggflac] ogg oga
 [flac] flac
 [opus] opus ogg oga
 [sndfile] wav aiff aif au snd paf iff svx sf voc w64 pvf xi htk caf sd2
 [audiofile] wav au aiff aif
 [dsf] dsf
 [hybrid_dsd] m4a
 [faad] aac
 [mpcdec] mpc
 [wavpack] wv
 [modplug] 669 amf ams dbm dfm dsm far it med mdl mod mtm mt2 okt s3m stm ult umx xm
 [mikmod] amf dsm far gdm imf it med mod mtm s3m stm stx ult uni xm
 [sidplay] sid mus str prg P00
 [wildmidi] mid
 [fluidsynth] mid
 [adplug] amd d00 hsc laa rad raw sa2
 [ffmpeg] 16sv 3g2 3gp 4xm 8svx aa3 aac ac3 adx afc aif aifc aiff al alaw amr anim apc ape asf atrac au aud avi avm2 avs bap bfi c93 cak cin cmv cpk daud dct divx dts dv dvd dxa eac3 film flac flc fli fll flx flv g726 gsm gxf iss m1v m2v m2t m2ts m4a m4b m4v mad mj2 mjpeg mjpg mka mkv mlp mm mmf mov mp+ mp1 mp2 mp3 mp4 mpc mpeg mpg mpga mpp mpu mve mvi mxf nc nsv nut nuv oga ogm ogv ogx oma ogg omg opus psp pva qcp qt r3d ra ram rl2 rm rmvb roq rpl rvc shn smk snd sol son spx str swf tak tgi tgq tgv thp ts tsp tta xa xvid uv uv2 vb vid vob voc vp6 vmd wav webm wma wmv wsaud wsvga wv wve rtp:// rtsp:// rtsps://
 [gme] ay gbs gym hes kss nsf nsfe rsn sap spc vgm vgz
 [dsdiff] dff
 [sacdiso] dat iso
 [dvdaiso] iso
 [pcm]

Filters:
 libsamplerate soxr

Tag plugins:
 id3tag

Output plugins:
 shout null fifo sndio pipe alsa ao openal pipewire pulse jack httpd snapcast recorder

Encoder plugins:
 null vorbis opus lame twolame wave flac shine

Archive plugins:
 [bz2] bz2
 [zzip] zip
 [iso] iso

Input plugins:
 file archive alsa qobuz curl ffmpeg nfs mms cdio_paranoia

Playlist plugins:
 extm3u m3u pls xspf asx rss soundcloud flac cue embcue

Protocols:
 file:// alsa:// cdda:// ftp:// ftps:// gopher:// hls+http:// hls+https:// http:// https:// mms:// mmsh:// mmst:// mmsu:// nfs:// qobuz:// rtmp:// rtmps:// rtmpt:// rtmpts:// rtp:// rtsp:// rtsps:// scp:// sftp:// smb:// srtp://

Other features:
 avahi dbus udisks epoll icu inotify ipv6 systemd tcp un
```

**Comment by MaxKellermann** [comment]
What files inside this ISO do you expect to see?

**Reply by c2b2pss** [reply]
Hi Max,

Thanks for the reply!!

I do not see the ISO image itself.  The ISO is an SACD ISO.  The ISO does not show up in my library list........don't fully understand how SACD ISO show under mpd.  What should I expect?

These ISOs work under foobar2000 in windows....so the ISOs are useable.  They dont show up in the list under mpd (via appropriate client).

If I use 'file' command on the ISO it says:

`/media/user/Elements/MyMusic/Master Superior Audiophile 2014.iso: UDF filesystem data (version 1.5) 'DISC1'`

So UDF format?

**Reply by MaxKellermann** [reply]
Then it's not an ISO image. The "ISO" in "ISO image" is an abbreviation for "ISO9660", which is a specific file system. UDF is a different file system. So what you have is not an ISO image, but a UDF image. MPD doesn't suport UDF images.

**Reply by c2b2pss** [reply]
Thanks again.  Sigh, that's what I figured.

So does mpd support SACD ISO in ISO9660 format?
I'm just curious to understand what I should expect......Are you aware of SACD ISO in the 9660 format? Any experience?

Regards

**Reply by MaxKellermann** [reply]
MPD supports ISO9660 images, but that's just a container for the actual music files. What kind of music files does such an image contain?
(We're back to my first question.)
I have no idea about SACD.

**Reply by c2b2pss** [reply]
DSD, DFF, or DTS. 2-channel and multichannel

From here:
http://forum.zidoo.tv/index.php?threads/tutorial-mch-music-sacd-options-and-iso-dsf-dff-file-formats.79385/


```
ISO/DSF/DFF files
Ripping an SACD results in an SACD ISO container with the DVD layer content only. The CD-layer is never part of it. Obviously that layer can be ripped using normal DAE (Digital Audio Extraction) tools if desired too. The SACD ISO contains all 2CH and/or MCH DSD tracks with or without DST compression as the final result. It is not possible to distinct the various options used without opening the container.

- Individual tracks can next be extracted from the full SACD ISO in DSF or DFF file format. Both are maintaining the original DSD 1-bit sampling format of the ISO. DSF (Sony) does not facilitate DST compression but DFF (Philips) does optionally. Having a DSF track you can be sure it is not compressed with DST anymore and only DSF supports tagged meta data making it the most popular extraction format.
- SACD ISO's with just 2CH tracks on them are mostly not DST compressed but when combined with MCH also those 2CH tracks are always compressed to reduce the total disc size. A MCH album would be 4x the size of an equivalent 2CH album without using DST compression!
- MCH albums have also 2CH tracks on them but there are a few very rare exceptions known (mostly classical concerts it seems). 2CH and MCH tracks hence need to be extracted individually. You can't see if a DSF or DFF file is MCH or 2CH and neither if a DFF file uses DST compression or not. If all Stereo (2CH) tracks aren't very short but the album is still around 1 Gbyte (ISO or DFF's) then it probably is DST compressed.
- It takes roughly 3x the total duration to extract a DST compressed SACD ISO to DSF or to DFF with DST decompression selected versus direct extraction to DFF with DST left untouched. MCH is 3x 2CH regarding size and it takes proportional resources to extract those tracks. So extracting a MCH DST track takes 9x the processing power compared to a 2CH uncompressed track! This is important to realize as it all needs to be done real time using media players in software which many can't cope with or not cope with well resulting in distorted playback if supported at all.
- DST is indeed very effective in size reduction (album reduced with roughly a factor 2.5). But even using DST some long MCH albums may still not fit on a single DVD layer. These then use both DVD layers for just the SACD tracks skipping the Hybrid CD compatibility layer as the consequence. Not a very common format but have a few of those myself (resulting ISO > 4.7 GByte).
```

It is easy to extract the audio files as dsd/dff etc and play as such, but with alot of such ISOs it gets tedious.

**Reply by MPconta** [reply]
I remember, some years ago there was a mpd-fork, which made it possible to recognize SACD-Images.
[https://gitlab.mel.vin/mirror/mpd-sacd](url)
I don't know, if this fork was merged into the actual MPD. Otherwise it won't help you a lot, because it seems to be unsupported.

**Reply by MaxKellermann** [reply]
It was not merged, because its code quality was not acceptable, and the author refused to improve it.
Though I don't know what was necessary to support SACD images - if it really contains DSF or DFF files, then MPD will be able to play it with the "iso9660" archive plugin, but only if it's ISO9660, and the problem is that what was called "ISO image" is not actually "ISO9660" but "UDF".

**Reply by c2b2pss** [reply]
I can report that I have a fully working and tested version of mpd source that handles SACD 'ISO' and DVD-A ISO ! My 'ISO' files play in mpd +ympd.  I tried multiple ISOs.  This is built from source.

The DVD-A MUST NOT be bluray DVD-A which also exists to play DTS-MA etc. I don't know yet the technical differences, but pure old sytle DVD-A audio (one can download samplers for free)

Re the 'iso' extension. As far as ripped SACD's go that is the default offering from rips, but a caveat must be issued on the conflicting meaning of ISO here.

Yes, I had to really scour the webs and really clean up the code since most copies of the author's source code failed with so many errors.  But is now builds on Ubunut 21 and plays through my DSD DAC. I am going to try to install it now on my headless server. I had to install and fidget so much that I want to make sure it plays on another machine !!

I would like to offer this source code to you, or if it is correct to fork it. I know nothing about hosting this stuff, so if you would like your advice, please let me know.

Regards!
